File: 2015/cs/AdventOfCode2015_CSharp/AdventOfCode2015.Core/Day01.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace AdventOfCode2015.Core
{
    public class Day01 : Solution2015
    {
        public override int Day => 1;
        public override string Title => "Not Quite Lisp";

        public override string Run(string input)
        {
            IEnumerable<int> parsed = parseInput(input);
            return FormatReport(SolvePartOne(parsed), SolvePartTwo(parsed));
        }

        public static IEnumerable<int> parseInput(string input)
            => input.Select(c => c == '(' ? 1 : -1);

        public int SolvePartOne(IEnumerable<int> input)
            => input.Sum();

        public int SolvePartTwo(IEnumerable<int> input)
        {
            return input.Accumulate((total, next) => total + next)
                .TakeWhile(floor => floor != -1)
                .Count() + 1;
        }
    }
}

File: 2015/java/src/main/java/com/robjwells/adventofcode2015/Day01.java
package com.robjwells.adventofcode2015;

import java.util.Arrays;

import static com.robjwells.adventofcode2015.Utils.accumulate;

public class Day01 extends Solution2015 {
    @Override
    public int getDay() {
        return 1;
    }

    @Override
    public String getTitle() {
        return "Not Quite Lisp";
    }

    @Override
    public String run(String input) {
        int[] parsed = parseInput(input);
        return formatReport(solvePartOne(parsed), solvePartTwo(parsed));
    }

    static int[] parseInput(String input) {
        return input.chars().map(c -> c == '(' ? 1 : -1).toArray();
    }

    static int solvePartOne(int[] deltas) {
        return Arrays.stream(deltas).sum();
    }

    static long solvePartTwo(int[] deltas) {
        return accumulate(Arrays.stream(deltas).iterator(), Math::addExact)
                .takeWhile(floor -> floor != -1)
                .count() + 1;
    }
}
